contracts/registry/src/registry.rs

```diff
@@ -425,4 +425,100 @@ impl SBTRegistry for Contract {
+    fn sbt_revoke_by_owner(&mut self, owner: AccountId, burn: bool) {
+        let issuer = env::predecessor_account_id();
+        let issuer_id = self.assert_issuer(&issuer);
+        let tokens_by_owner =
+            self.sbt_tokens_by_owner(owner.clone(), Some(issuer.clone()), None, None, Some(true));
+        require!(
+            !tokens_by_owner.is_empty(),
+            "no tokens to be revoked for the given owner"
+        );
+        let tokens: Vec<OwnedToken> = tokens_by_owner[0].1.iter().map(|t| t.clone()).collect();
+        let token_ids: Vec<u64> = tokens.iter().map(|s| s.token).collect();
+        if burn == true {
+            let mut revoked_per_class: HashMap<u64, u64> = HashMap::new();
+            let tokens_burned: u64 = tokens.len().try_into().unwrap();
+            for token in &tokens {
+                // update balances
+                let class_id = token.metadata.class;
+                let balance_key = &BalanceKey {
+                    issuer_id,
+                    owner: owner.clone(),
+                    class_id: token.metadata.class,
+                };
+                self.balances.remove(balance_key);
+
+                // collect the info about the tokens revoked per class
+                // to update the balance accordingly
+                revoked_per_class
+                    .entry(class_id)
+                    .and_modify(|key_value| *key_value += 1)
+                    .or_insert(1);
+
+                // remove from issuer_tokens
+                self.issuer_tokens.remove(&IssuerTokenId {
+                    issuer_id,
+                    token: token.token,
+                });
+            }
+
+            // update supply by owner
+            let old_supply = self
+                .supply_by_owner
+                .get(&(owner.clone(), issuer_id))
+                .unwrap();
+            self.supply_by_owner.insert(
+                &(owner.clone(), issuer_id),
+                &(old_supply - &(tokens.len() as u64)),
+            );
+
+            // update supply by class
+            for (class_id, tokens_revoked) in revoked_per_class {
+                let old_supply = self.supply_by_class.get(&(issuer_id, class_id)).unwrap();
+                self.supply_by_class
+                    .insert(&(issuer_id, class_id), &(&old_supply - &tokens_revoked));
+            }
+
+            // update supply by issuer
+            let supply_by_issuer = self.supply_by_issuer.get(&(issuer_id)).unwrap_or(0);
+            self.supply_by_issuer
+                .insert(&(issuer_id), &(supply_by_issuer - tokens_burned));
+
+            // emit event
+            SbtTokensEvent {
+                issuer: issuer.clone(),
+                tokens: token_ids.clone(),
+            }
+            .emit_burn();
+        } else {
+            let current_timestamp = env::block_timestamp();
+            // revoke
+            for mut token in tokens.clone() {
```

Comment:
It is being cloned here because we are assigning to it inside of the loop. Rust does not allow us to change the values in the struct we are iterating over